_posts/2024-04-20-websockets2.md

```diff
@@ -0,0 +1,1861 @@
+
+### 6.3. The transferUserToRoom() Function
+This is a private function, that transfers the user to a new room:
+
+```scala
+object ChatProtocol {
+  ...
+  private def transferUserToRoom[F[_]: Monad: UUIDGen: Concurrent](
+      redisP: RedisProtocol[F],
+      postgresP: PostgresProtocol[F],
+      user: User,
+      room: Room
+  ): F[List[OutputMessage]] =
+    val leaveMessages = removeFromCurrentRoom(redisP, postgresP, user)
+    val enterMessages = addToRoom(redisP, postgresP, user, room)
+    for {
+      leave <- leaveMessages
+      enter <- enterMessages
+    } yield leave ++ enter
+}
+```
+To do this we remove the user from the current room by calling `removeFromCurrentRoom()`, then add the user to the requested room by calling `addToRoom(redisP, postgresP, user, room)`. This happens both in Redis and Postgres to keep everything in sync.
+
+### 6.4. The removeFromCurrentRoom() Function
+To remove a user from the current room, we'll need to remove the user from the Redis `room:<roomid>` set and remove the entry from the `userroomid` hash. 
+
+Note that in Redis when the last member is deleted from a set, the entire set is deleted, therefore if this occurs, we'll also need to update the `rooms` hash:
+
+```scala
+new RedisProtocol[F] {
+  ...
+  override def removeUserFromRoom(
+      roomid: RoomId,
+      userid: UserId
+  ): F[Unit] = {
+    redis.sRem(s"room:${roomid.id.toString}", userid.id.toString).void
+  }
+}
+```
+The `removeUserFromRoom()` uses the `sRem` function to delete the user from the `room:<roomid>` set. It requires the room name and the user id:
+
+```scala
+new RedisProtocol[F] {
+  ...
+  override def deleteUserRoomMapping(userid: UserId): F[Unit] = {
+    redis.hDel("userroomid", userid.id.toString).void
+  }
+
+  override def deleteRoom(roomid: RoomId): F[Long] = {
+    redis.hDel("rooms", roomid.id.toString)
+  }
+}
+```
+The `deleteUserRoomMapping()` and `deleteRoom()`  uses the `hDel` function to delete an entry from the `userroomid` and `rooms` hashes, and only requires the user id and room id respectively.
+
+Since Redis automatically deletes empty sets, we'll have to check if the `room:<roomid>` set still exists:
+
+```scala
+new RedisProtocol[F] {
+  ...
+  override def roomExists(roomid: RoomId): F[Boolean] = {
+    redis.exists(s"room:${roomid.id.toString}")
+  }
+}
+```
+We use the `exists` function which checks if a key exists within redis.
+
+Finally, we can implement the removeFromCurrentRoom() function:
+
+```scala
+object ChatProtocol {
+  ...
+  private def removeFromCurrentRoom[F[_]: Monad: UUIDGen](
+        redisP: RedisProtocol[F],
+        postgresP: PostgresProtocol[F],
+        user: User
+    ): F[List[OutputMessage]] = {
+      redisP.getUsersRoomId(user).flatMap {
+        case Some(roomid) =>
+          for
+            _ <- redisP.removeUserFromRoom(roomid, user.id)
+            _ <- redisP.roomExists(roomid).flatMap { b =>
```

Comment:
do we want to delete the room in case the last user exits?

Reply:
Yes, once the last user exits, that set is automatically deleted in Redis, so we do the same in Postgres.